File: ontrack-ui/src/main/java/net/nemerosa/ontrack/boot/ProjectSearchProvider.java
package net.nemerosa.ontrack.boot;

import net.nemerosa.ontrack.boot.ui.ProjectController;
import net.nemerosa.ontrack.model.structure.NameDescription;
import net.nemerosa.ontrack.model.structure.Project;
import net.nemerosa.ontrack.model.structure.SearchResult;
import net.nemerosa.ontrack.model.structure.StructureService;
import net.nemerosa.ontrack.ui.controller.URIBuilder;
import net.nemerosa.ontrack.ui.support.AbstractSearchProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.mvc.method.annotation.MvcUriComponentsBuilder;

import java.util.Collection;
import java.util.Collections;
import java.util.regex.Pattern;

@Component
public class ProjectSearchProvider extends AbstractSearchProvider {

    private final StructureService structureService;

    @Autowired
    public ProjectSearchProvider(URIBuilder uriBuilder, StructureService structureService) {
        super(uriBuilder);
        this.structureService = structureService;
    }

    @Override
    public boolean isTokenSearchable(String token) {
        return Pattern.matches(NameDescription.NAME, token);
    }

    @Override
    public Collection<SearchResult> search(String token) {
        Project project = structureService.findProjectByName(token);
        if (project != null) {
            return Collections.singletonList(
                    new SearchResult(
                            project.getName(),
                            String.format("%s project", project.getName()),
                            uri(MvcUriComponentsBuilder.on(ProjectController.class).getProject(project.getId())),
                            String.format("/project/%d", project.id()),
                            100
                    )
            );
        } else {
            return Collections.emptyList();
        }
    }
}

File: ontrack-ui/src/main/java/net/nemerosa/ontrack/boot/ProjectSearchProvider.kt
package net.nemerosa.ontrack.boot

import co.elastic.clients.elasticsearch._types.query_dsl.Query
import co.elastic.clients.elasticsearch._types.query_dsl.TextQueryType
import co.elastic.clients.elasticsearch.indices.CreateIndexRequest
import co.elastic.clients.util.ObjectBuilder
import com.fasterxml.jackson.databind.JsonNode
import net.nemerosa.ontrack.extension.support.CoreExtensionFeature
import net.nemerosa.ontrack.model.events.Event
import net.nemerosa.ontrack.model.events.EventFactory
import net.nemerosa.ontrack.model.events.EventListener
import net.nemerosa.ontrack.model.structure.*
import org.springframework.stereotype.Component

/**
 * Minimum length of a token to compare with project names by content.
 */
const val PROJECT_SEARCH_PROVIDER_TOKEN_MIN_LENGTH = 3

/**
 * Projects search index name
 */
const val PROJECT_SEARCH_INDEX = "projects"

/**
 * Search result type
 */
const val PROJECT_SEARCH_RESULT_TYPE = "project"

/**
 * Matches against projects either by name or by content if the
 * search token is at least [PROJECT_SEARCH_PROVIDER_TOKEN_MIN_LENGTH] characters long.
 */
@Component
class ProjectSearchProvider(
    private val structureService: StructureService,
    private val searchIndexService: SearchIndexService
) : SearchIndexer<ProjectSearchItem>, EventListener {

    override val searchResultType = SearchResultType(
        feature = CoreExtensionFeature.INSTANCE.featureDescription,
        id = PROJECT_SEARCH_RESULT_TYPE,
        name = "Project",
        description = "Project name in Ontrack",
        order = SearchResultType.ORDER_PROJECT,
    )

    override val indexerName: String = "Projects"

    override val indexName: String = PROJECT_SEARCH_INDEX

    override fun initIndex(builder: CreateIndexRequest.Builder): CreateIndexRequest.Builder {
        return builder.run {
            autoCompleteSettings()
        }.run {
            mappings { mappings ->
                mappings
                    .autoCompleteText(ProjectSearchItem::name)
                    .text(ProjectSearchItem::description)
            }
        }
    }

    override fun buildQuery(
        q: Query.Builder,
        token: String
    ): ObjectBuilder<Query> {
        return q.multiMatch { m ->
            m.query(token)
                .type(TextQueryType.BestFields)
                .fields(
                    ProjectSearchItem::name to 3.0,
                    ProjectSearchItem::description to null,
                )
        }
    }

    override fun indexAll(processor: (ProjectSearchItem) -> Unit) {
        structureService.projectList.forEach { project ->
            processor(project.asSearchItem())
        }
    }

    override fun toSearchResult(id: String, score: Double, source: JsonNode): SearchResult? {
        return structureService.findProjectByID(ID.of(id.toInt()))?.run {
            SearchResult(
                title = entityDisplayName,
                description = description ?: "",
                accuracy = score,
                type = searchResultType,
                data = mapOf(
                    SearchResult.SEARCH_RESULT_PROJECT to this,
                )
            )
        }
    }

    override fun onEvent(event: Event) {
        when (event.eventType) {
            EventFactory.NEW_PROJECT -> {
                val project = event.getEntity<Project>(ProjectEntityType.PROJECT)
                searchIndexService.createSearchIndex(this, project.asSearchItem())
            }

            EventFactory.UPDATE_PROJECT -> {
                val project = event.getEntity<Project>(ProjectEntityType.PROJECT)
                searchIndexService.updateSearchIndex(this, project.asSearchItem())
            }

            EventFactory.DELETE_PROJECT -> {
                val projectId = event.getIntValue("PROJECT_ID")
                searchIndexService.deleteSearchIndex(this, projectId)
            }
        }
    }

    private fun Project.asSearchItem() = ProjectSearchItem(this)

}

class ProjectSearchItem(
    override val id: String,
    val name: String,
    val description: String
) : SearchItem {
    constructor(project: Project) : this(
        id = project.id.toString(),
        name = project.name,
        description = project.description ?: ""
    )

    override val fields: Map<String, Any> = mapOf(
        "name" to name,
        "description" to description
    )

}
